README.md
# prefect-dbtcloud

## Welcome!

Collection of tasks to interact with dbt Cloud

## Getting Started

### Python setup

Requires an installation of Python 3.7+.

We recommend using a Python virtual environment manager such as pipenv, conda or virtualenv.

These tasks are designed to work with Prefect 2.0. For more information about how to use Prefect, please refer to the [Prefect documentation](https://orion-docs.prefect.io/).

### Installation

Install `prefect-dbtcloud` with `pip`:

```bash
pip install prefect-dbtcloud
```

### Write and run a flow

```python
from prefect import flow
from prefect_dbtcloud.tasks import (
    create_job,
    run_job
)

# Create a new job
@flow
def create_new_job():
    create_job(
        account_id=111,                   #  Your dbt Cloud Account ID.
        project_id=222,                   #  The ID of the project where to create the job.
        environment_id=333,               #  The ID of the environment to use to run the job.
        name="new-job",                   #  The name of the job.
        execute_steps=["dbt run"],        #  The list of dbt commands the job will execute.
        generate_docs=True                #  Whether dbt should generate docs or not.
    )

# Trigger job run
@flow
def trigger_dbt_cloud_job_run():
    run_job(
        cause="The cause",                 #  A string describing why you're triggering the job.
        account_id=111,                    #  Your dbt Cloud Account ID.
        job_id=678,                        #  The ID of the Job to run.
        token="The secret token",          #  The API token to use to authenticate on dbt Cloud.
        wait_for_job_run_completion=False  #  Whether to wait for job completion or not.
    )

create_new_job()
trigger_dbt_cloud_job_run()
```

## Resources

If you encounter any bugs while using `prefect-dbtcloud`, feel free to open an issue in the [prefect-dbtcloud](https://github.com/AlessandroLollo/prefect-dbtcloud) repository.

If you have any questions or issues while using `prefect-dbtcloud`, you can find help in either the [Prefect Discourse forum](https://discourse.prefect.io/) or the [Prefect Slack community](https://prefect.io/slack).

## Development

If you'd like to install a version of `prefect-dbtcloud` for development, clone the repository and perform an editable install with `pip`:

```bash
git clone https://github.com/AlessandroLollo/prefect-dbtcloud.git

cd prefect-dbtcloud/

pip install -e ".[dev]"

# Install linting pre-commit hooks
pre-commit install
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## 0.1.0

Released on ????? ?th, 20??.

### Added

- `task_name` task - [#1](https://github.com/AlessandroLollo/prefect-dbtcloud/pull/1)


tests/test_flows.py
from prefect_dbtcloud.flows import hello_and_goodbye


def test_hello_and_goodbye_flow():
    flow_state = hello_and_goodbye()
    assert flow_state.is_completed


tests/test_tasks.py
import os
from unittest import mock

import pytest
import responses
from prefect import flow
from responses import matchers

from prefect_dbtcloud.exceptions import (
    DbtCloudConfigurationException,
    DbtCloudCreateJobFailed,
    DbtCloudRunCanceled,
    DbtCloudRunFailed,
    DbtCloudRunTimedOut,
    GetDbtCloudRunFailed,
    TriggerDbtCloudRunFailed,
)
from prefect_dbtcloud.tasks import create_job, run_job
from prefect_dbtcloud.utils import dbtCloudClient


def test_run_without_account_id_raises():
    @flow
    def test_flow():
        return run_job(cause="cause")

    msg_match = "dbt Cloud Account ID cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_run_without_job_id_raises():
    @flow
    def test_flow():
        return run_job(cause="cause", account_id=123)

    msg_match = "dbt Cloud Job ID cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_run_without_token_raises():
    @flow
    def test_flow():
        return run_job(cause="cause", account_id=123, job_id=123)

    msg_match = "dbt Cloud token cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_run_with_cause_none_raises():
    @flow
    def test_flow():
        return run_job(cause=None, account_id=123, job_id=123, token="abc")

    msg_match = "Cause cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


@responses.activate
def test_run_job_failed_raises():
    account_id = 123
    job_id = 123
    url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"

    responses.add(
        responses.POST,
        url,
        status=123,
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(cause="abc", account_id=account_id, job_id=job_id, token="abc")

    with pytest.raises(TriggerDbtCloudRunFailed):
        test_flow().result().result()


@responses.activate
def test_run_job_with_wait_failed_raises():
    account_id = 123
    job_id = 123
    trigger_url = (
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )
    get_run_url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/"

    responses.add(
        responses.POST,
        trigger_url,
        status=200,
        json={"data": {"id": 123}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=123,
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="abc",
            account_id=account_id,
            job_id=job_id,
            token="abc",
            wait_for_job_run_completion=True,
        )

    with pytest.raises(GetDbtCloudRunFailed):
        test_flow().result().result()


@responses.activate
def test_run_job_with_wait_status_20_raises():
    account_id = 123
    job_id = 123
    trigger_url = (
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )
    get_run_url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/"

    responses.add(
        responses.POST,
        trigger_url,
        status=200,
        json={"data": {"id": 123}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=200,
        json={"data": {"status": 20, "finished_at": "2019-08-24T14:15:22Z"}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="abc",
            account_id=account_id,
            job_id=job_id,
            token="abc",
            wait_for_job_run_completion=True,
        )

    with pytest.raises(DbtCloudRunFailed):
        test_flow().result().result()


@responses.activate
def test_run_job_with_wait_status_30_raises():
    account_id = 123
    job_id = 123
    trigger_url = (
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )
    get_run_url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/"

    responses.add(
        responses.POST,
        trigger_url,
        status=200,
        json={"data": {"id": 123}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=200,
        json={"data": {"status": 30, "finished_at": "2019-08-24T14:15:22Z"}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="abc",
            account_id=account_id,
            job_id=job_id,
            token="abc",
            wait_for_job_run_completion=True,
        )

    with pytest.raises(DbtCloudRunCanceled):
        test_flow().result().result()


@responses.activate
def test_run_job_timeout_expires_raises():
    account_id = 123
    job_id = 123
    trigger_url = (
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )
    get_run_url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/"

    responses.add(
        responses.POST,
        trigger_url,
        status=200,
        json={"data": {"id": 123}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=200,
        json={"data": {"finished_at": None}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="abc",
            account_id=account_id,
            job_id=job_id,
            token="abc",
            wait_for_job_run_completion=True,
            max_wait_time=5,
        )

    with pytest.raises(DbtCloudRunTimedOut):
        test_flow().result().result()


@responses.activate
def test_run_job_trigger_job():
    account_id = 123
    job_id = 123
    url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"

    responses.add(
        responses.POST,
        url,
        json={"data": {"foo": "bar"}},
        status=200,
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(cause="abc", account_id=account_id, job_id=job_id, token="abc")

    response = test_flow().result().result()

    assert response == {"foo": "bar"}


@responses.activate
def test_run_job_trigger_job_with_custom_domain():
    account_id = 123
    job_id = 123
    api_domain = "cloud.corp.getdbt.com"
    url = f"https://{api_domain}/api/v2/accounts/{account_id}/jobs/{job_id}/run/"

    responses.add(
        responses.POST,
        url,
        json={"data": {"foo": "bar"}},
        status=200,
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="abc",
            account_id=account_id,
            job_id=job_id,
            token="abc",
            api_domain=api_domain,
        )

    response = test_flow().result().result()

    assert response == {"foo": "bar"}


@responses.activate
def test_dbt_cloud_run_job_trigger_job_with_wait():
    account_id = 1234
    job_id = 1234

    trigger_url = (
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )
    get_run_url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/"

    responses.add(
        responses.POST,
        trigger_url,
        status=200,
        json={"data": {"id": 123}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=200,
        json={"data": {"id": 1, "status": 10, "finished_at": "2019-08-24T14:15:22Z"}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/123/artifacts/",
        status=200,
        json={"data": ["manifest.json", "run_results.json", "catalog.json"]},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="foo",
            account_id=account_id,
            job_id=job_id,
            token="foo",
            wait_for_job_run_completion=True,
        )

    r = test_flow().result().result()

    assert r == {
        "id": 1,
        "status": 10,
        "finished_at": "2019-08-24T14:15:22Z",
        "artifact_urls": [
            f"{get_run_url}artifacts/manifest.json",
            f"{get_run_url}artifacts/run_results.json",
            f"{get_run_url}artifacts/catalog.json",
        ],
    }


@responses.activate
def test_dbt_cloud_run_job_trigger_job_with_wait_custom():
    account_id = 1234
    job_id = 1234

    trigger_run_url = (
        f"https://cloud.corp.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/"
    )

    get_run_url = f"https://cloud.corp.getdbt.com/api/v2/accounts/{account_id}/runs/1/"

    responses.add(
        responses.POST,
        trigger_run_url,
        status=200,
        json={"data": {"id": 1}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        get_run_url,
        status=200,
        json={"data": {"id": 1, "status": 10, "finished_at": "2019-08-24T14:15:22Z"}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        f"https://cloud.corp.getdbt.com/api/v2/accounts/{account_id}/runs/1/artifacts/",
        status=200,
        json={"data": ["manifest.json", "run_results.json", "catalog.json"]},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="foo",
            account_id=account_id,
            job_id=job_id,
            token="foo",
            wait_for_job_run_completion=True,
            api_domain="cloud.corp.getdbt.com",
        )

    r = test_flow().result().result()

    assert r == {
        "id": 1,
        "status": 10,
        "finished_at": "2019-08-24T14:15:22Z",
        "artifact_urls": [
            f"{get_run_url}artifacts/manifest.json",
            f"{get_run_url}artifacts/run_results.json",
            f"{get_run_url}artifacts/catalog.json",
        ],
    }


@responses.activate
def test_dbt_cloud_run_job_trigger_job_with_fail_on_artifacts():
    account_id = 1234
    job_id = 1234

    responses.add(
        responses.POST,
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/{job_id}/run/",
        status=200,
        json={"data": {"id": 1}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/1/",
        status=200,
        json={"data": {"id": 1, "status": 10, "finished_at": "2019-08-24T14:15:22Z"}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    responses.add(
        responses.GET,
        f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/runs/1/artifacts/",
        status=123,
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    @flow
    def test_flow():
        return run_job(
            cause="foo",
            account_id=account_id,
            job_id=job_id,
            token="foo",
            wait_for_job_run_completion=True,
            api_domain="cloud.getdbt.com",
        )

    r = test_flow().result().result()

    assert r == {
        "id": 1,
        "status": 10,
        "finished_at": "2019-08-24T14:15:22Z",
        "artifact_urls": [],
    }


@responses.activate
@mock.patch.dict(os.environ, {"ACCT_ID": "123", "JOB": "123", "TKN": "abc"})
def test_run_job_with_env_vars():

    responses.add(
        responses.POST,
        "https://cloud.getdbt.com/api/v2/accounts/123/jobs/123/run/",
        status=200,
        json={"data": {"id": 1}},
        match=[matchers.header_matcher(dbtCloudClient.get_agent_header())],
    )

    account_id_env_var_name = "ACCT_ID"
    job_id_env_var_name = "JOB"
    token_env_var_name = "TKN"

    @flow
    def test_flow():
        return run_job(
            account_id_env_var_name=account_id_env_var_name,
            job_id_env_var_name=job_id_env_var_name,
            token_env_var_name=token_env_var_name,
            wait_for_job_run_completion=False,
            cause="test",
        )

    r = test_flow().result().result()

    assert r == {"id": 1}


def test_create_without_account_id_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=None, environment_id=123, project_id=123, name="abc"
        )

    msg_match = "dbt Cloud Account ID cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_create_without_token_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=None,
            environment_id=123,
            project_id=123,
            name="abc",
            account_id=123,
        )

    msg_match = "dbt Cloud token cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_create_without_execute_steps_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=None,
            environment_id=123,
            project_id=123,
            name="abc",
            account_id=123,
            token="tkn",
        )

    msg_match = "Steps executed by dbt Cloud job cannot be None or empty."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_create_without_project_id_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=["dbt run"],
            environment_id=123,
            project_id=None,
            name="abc",
            account_id=123,
            token="tkn",
        )

    msg_match = "dbt Cloud Project ID cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_create_without_environment_id_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=["dbt run"],
            environment_id=None,
            project_id=123,
            name="abc",
            account_id=123,
            token="tkn",
        )

    msg_match = "dbt Cloud Environment ID cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


def test_create_without_name_raises():
    @flow
    def test_flow():
        return create_job(
            execute_steps=["dbt run"],
            environment_id=123,
            project_id=123,
            name=None,
            account_id=123,
            token="tkn",
        )

    msg_match = "dbt Cloud Job name cannot be None."
    with pytest.raises(DbtCloudConfigurationException, match=msg_match):
        test_flow().result().result()


@responses.activate
def test_create_with_failure_raises():
    account_id = 123
    url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/"

    responses.add(responses.POST, url, status=123)

    @flow
    def test_flow():
        return create_job(
            account_id=account_id,
            environment_id=123,
            project_id=123,
            name="abc",
            token="tkn",
            execute_steps=["dbt run"],
        )

    with pytest.raises(DbtCloudCreateJobFailed):
        test_flow().result().result()


@responses.activate
def test_create_succeed():
    account_id = 123
    name = "test-job"
    job_id = 123
    environment_id = 123
    project_id = 123
    dbt_version = "1.0.0"
    execute_steps = ["dbt run", "dbt test"]
    triggers = {"github_webhook": False}
    settings = {"threads": 8}
    generate_docs = True
    schedule = {"cron": "cron-string"}
    url = f"https://cloud.getdbt.com/api/v2/accounts/{account_id}/jobs/"

    responses.add(
        responses.POST,
        url,
        status=200,
        json={
            "data": {
                "account_id": account_id,
                "job_id": job_id,
                "environment_id": environment_id,
                "project_id": project_id,
                "name": name,
                "execute_steps": execute_steps,
                "triggers": triggers,
                "settings": settings,
                "dbt_version": dbt_version,
                "generate_docs": generate_docs,
                "schedule": schedule,
            }
        },
    )

    @flow
    def test_flow():
        return create_job(
            account_id=account_id,
            project_id=project_id,
            environment_id=environment_id,
            name=name,
            dbt_version=dbt_version,
            execute_steps=execute_steps,
            triggers=triggers,
            settings=settings,
            generate_docs=generate_docs,
            schedule=schedule,
            token="tkn",
        )

    r = test_flow().result().result()

    assert r["account_id"] == account_id
    assert r["project_id"] == project_id
    assert r["environment_id"] == environment_id
    assert r["name"] == name
    assert r["dbt_version"] == dbt_version
    assert r["triggers"] == triggers
    assert r["settings"] == settings
    assert r["generate_docs"] == generate_docs
    assert r["schedule"] == schedule


prefect_dbtcloud/exceptions.py
"""
Exceptions to be raised in case of issues or unexpected behaviours
of dbt Cloud.
"""


class DbtCloudBaseException(Exception):
    """Base exception for all dbt Cloud errors"""

    pass


class DbtCloudConfigurationException(Exception):
    """Raise when dbt Cloud client is misconfigured"""

    pass


class DbtCloudCreateJobFailed(DbtCloudBaseException):
    """Raise when job creation fails"""

    pass


class TriggerDbtCloudRunFailed(DbtCloudBaseException):
    """Raised when triggering a dbt job run fails"""

    pass


class GetDbtCloudRunFailed(DbtCloudBaseException):
    """Raised when details for a dbt Cloud job run cannot be retrieved"""

    pass


class DbtCloudRunFailed(DbtCloudBaseException):
    """Raised when a dbt Cloud run fails"""

    pass


class DbtCloudRunCanceled(DbtCloudBaseException):
    """Raised when a dbt Cloud run has been canceled before completion"""

    pass


class DbtCloudRunTimedOut(DbtCloudBaseException):
    """Raised when a dbt Cloud run does not complete in the provided time"""

    pass


class DbtCloudListArtifactsFailed(DbtCloudBaseException):
    """Raised when dbt Cloud artifacts cannot be listed"""

    pass


prefect_dbtcloud/utils.py
"""
Utilities and client to interact with dbt Cloud.
"""
from time import sleep
from typing import Dict, List, Optional, Tuple

import prefect
from requests import Session

from prefect_dbtcloud.exceptions import (
    DbtCloudCreateJobFailed,
    DbtCloudListArtifactsFailed,
    DbtCloudRunCanceled,
    DbtCloudRunFailed,
    DbtCloudRunTimedOut,
    GetDbtCloudRunFailed,
    TriggerDbtCloudRunFailed,
)


class dbtCloudClient:
    """
    A client that exposes methods to call dbt Cloud APIs
    to trigger job runs and retrieve job run information.

    """

    # dbt Cloud Create Job API
    # https://docs.getdbt.com/dbt-cloud/api-v2#operation/createJob
    __DBT_CLOUD_CREATE_JOB_API_ENDPOINT_V2 = (
        "https://{apiDomain}/api/v2/accounts/{accountId}/jobs/"
    )

    # dbt Cloud Trigger Job API
    # https://docs.getdbt.com/dbt-cloud/api-v2#operation/triggerRun
    __DBT_CLOUD_TRIGGER_JOB_API_ENDPOINT_V2 = (
        "https://{apiDomain}/api/v2/accounts/{accountId}/jobs/{jobId}/run/"
    )

    # dbt Cloud Get Run API
    # https://docs.getdbt.com/dbt-cloud/api-v2#operation/getRunById
    __DBT_CLOUD_GET_RUN_API_ENDPOINT_V2 = (
        "https://{apiDomain}/api/v2/accounts/{accountId}/runs/{runId}/"
    )

    # dbt Cloud List Run Artifacts API
    # https://docs.getdbt.com/dbt-cloud/api-v2#operation/listArtifactsByRunId
    __DBT_CLOUD_LIST_RUN_ARTIFACTS_ENDPOINT_V2 = (
        "https://{apiDomain}/api/v2/accounts/{accountId}/runs/{runId}/artifacts/"
    )

    # dbt Cloud Get Run Artifact API
    # https://docs.getdbt.com/dbt-cloud/api-v2#operation/getArtifactsByRunId
    __DBT_CLOUD_GET_RUN_ARTIFACT_ENDPOINT_V2 = (
        "https://{apiDomain}/api/v2/accounts/{accountId}/runs/{runId}/artifacts/{path}"
    )

    __USER_AGENT_HEADER = {"user-agent": f"prefect-{prefect.__version__}"}

    def __init__(
        self,
        account_id: int,
        token: str,
        api_domain: Optional[str] = "cloud.getdbt.com",
    ) -> "dbtCloudClient":
        """
        Build a `dbtCloudClient` object.

        Args:
            account_id: The identifier of the dbt Cloud account.
            token: The API token to use to authenticate on dbt Cloud.
            api_domain: The URL of the dbt Cloud account.
                Defaults to "cloud.getdbt.com"

        Returns:
            A `dbtCloudClient` configured to interact with the
            specified dbt Cloud account
        """
        self.account_id = account_id
        self.api_domain = api_domain
        self.token = token
        self.session = Session()
        self.session.headers = {
            "Authorization": f"Bearer {self.token}",
            **self.__USER_AGENT_HEADER,
        }

    @classmethod
    def get_agent_header(cls) -> Dict:
        """
        Return the user agent header
        the client uses when performing API requests
        to dbt Cloud.

        Returns:
            User agent header
        """
        return cls.__USER_AGENT_HEADER

    def dbt_cloud_create_job_api_endpoint_v2(self) -> str:
        """
        Return the URL of the Create Job API.

        Returns:
            The URL of the Create Job API.
        """
        return self.__DBT_CLOUD_CREATE_JOB_API_ENDPOINT_V2.format(
            apiDomain=self.api_domain, accountId=self.account_id
        )

    def dbt_cloud_trigger_job_api_endpoint_v2(self, job_id: int) -> str:
        """
        Return the URL of the Trigger Job API.

        Args:
            job_id: The identifier of the job to trigger.

        Returns:
            The URL of the Trigger Job API for the given `job_id`.
        """
        return self.__DBT_CLOUD_TRIGGER_JOB_API_ENDPOINT_V2.format(
            apiDomain=self.api_domain, accountId=self.account_id, jobId=job_id
        )

    def dbt_cloud_get_run_api_endpoint_v2(self, run_id: int) -> str:
        """
        Return the URL of the Get Run API.

        Args:
            run_id: the identifier of the job run to retrieve.

        Returns:
            The URL of the Get Run API for the given `run_id`.
        """
        return self.__DBT_CLOUD_GET_RUN_API_ENDPOINT_V2.format(
            apiDomain=self.api_domain, accountId=self.account_id, runId=run_id
        )

    def dbt_cloud_list_run_artifacts_endpoint_v2(self, run_id: int) -> str:
        """
        Return the URL of the List Run Artifacts API.

        Args:
            run_id: the identifier of the job run to retrieve.

        Returns:
            The URL of the List Run Artifacts API for the given `run_id`.
        """
        return self.__DBT_CLOUD_LIST_RUN_ARTIFACTS_ENDPOINT_V2.format(
            apiDomain=self.api_domain, accountId=self.account_id, runId=run_id
        )

    def dbt_cloud_get_run_artifact_endpoint_v2(self, run_id: int, path: str) -> str:
        """
        Return the URL of the Get Run Artifact API.

        Args:
            run_id: the identifier of the job run to retrieve.

        Returns:
            The URL of the Get Run Artifact API for the given `run_id`.
        """
        return self.__DBT_CLOUD_GET_RUN_ARTIFACT_ENDPOINT_V2.format(
            apiDomain=self.api_domain,
            accountId=self.account_id,
            runId=run_id,
            path=path,
        )

    def create_job(
        self,
        project_id: int,
        environment_id: int,
        name: str,
        execute_steps: List[str],
        dbt_version: Optional[str] = None,
        triggers: Optional[Dict] = None,
        settings: Optional[Dict] = None,
        generate_docs: Optional[bool] = False,
        schedule: Optional[Dict] = None,
    ):
        """
        Create a new dbt Cloud job.

        Args:
            project_id: the ID of the project where the
                job will be created.
            environment_id: the ID of the environment that will
                be used to run the job.
            name: the name of the job.
            execute_steps: The list of dbt commands the job
                will execute when triggered.
            dbt_version: the dbt version to use to run the job.
                If provided, it will override the one defined
                in the environment.
            triggers: an object describing which trigger types
                will be enabled for the job.
            settings: an object containing settings to be applied
                to the job when running.
            generate_docs: whether to run `dbt docs generate` or not,
                after the job has been executed.
            schedule: an object that contains the run schedule
                specification for the job.

        Returns:
            The create job result, namely the `data` key in the API response.

        Raises:
            `DbtCloudCreateJobFailed`: when the response code != 200
        """
        url = self.dbt_cloud_create_job_api_endpoint_v2()

        data = {}
        data["project_id"] = project_id
        data["environment_id"] = environment_id
        data["name"] = name
        data["execute_steps"] = execute_steps

        if dbt_version:
            data["dbt_version"] = dbt_version

        if triggers:
            data["triggers"] = triggers

        if settings:
            data["settings"] = settings

        if generate_docs:
            data["generate_docs"] = generate_docs

        if schedule:
            data["schedule"] = schedule

        with self.session.post(url, data=data) as create_job_request:

            if create_job_request.status_code != 200:
                raise DbtCloudCreateJobFailed(create_job_request.reason)

        return create_job_request.json()["data"]

    def trigger_job_run(
        self,
        job_id: int,
        cause: str,
        additional_args: Optional[Dict] = None,
    ) -> dict:
        """
        Trigger a dbt Cloud job run

        Args:
            job_id: dbt Cloud job ID
            cause: the reason describing why the job run is being triggered
            additional_args: additional information to pass to the Trigger Job Run API

        Returns:
            The trigger run result, namely the `data` key in the API response

        Raises:
            `TriggerDbtCloudRunFailed`: when the response code is != 200
        """
        data = additional_args if additional_args else {}
        data["cause"] = cause

        url = self.dbt_cloud_trigger_job_api_endpoint_v2(job_id=job_id)
        with self.session.post(url, data=data) as trigger_request:

            if trigger_request.status_code != 200:
                raise TriggerDbtCloudRunFailed(trigger_request.reason)

        return trigger_request.json()["data"]

    def wait_for_job_run(self, run_id: int, max_wait_time: int = None) -> Dict:
        """
        Get a dbt Cloud job run.
        Please note that this function will fail if any call to dbt Cloud APIs fail.

        Args:
            run_id: dbt Cloud job run ID
            max_wait_time: the number od seconds to wait for the job to complete

        Returns:
            The job run result, namely the "data" key in the API response

        Raises:
            `DbtCloudRunFailed`: if "finished_at" is not None
                and the result status == 20
            `DbtCloudRunCanceled`: if "finished_at" is not None
                and the result status == 30
            `DbtCloudRunTimedOut`: if run does not finish
                before provided max_wait_time
        """
        wait_time_between_api_calls = 10
        elapsed_wait_time = 0
        url = self.dbt_cloud_get_run_api_endpoint_v2(run_id=run_id)

        while not max_wait_time or elapsed_wait_time <= max_wait_time:
            with self.session.get(url) as get_run_request:

                if get_run_request.status_code != 200:
                    raise GetDbtCloudRunFailed(get_run_request.reason)

            result = get_run_request.json()["data"]

            if result["finished_at"]:
                if result["status"] == 10:
                    return result
                elif result["status"] == 20:
                    raise DbtCloudRunFailed(f"Job run with ID: {run_id} failed.")
                elif result["status"] == 30:
                    raise DbtCloudRunCanceled(f"Job run with ID: {run_id} cancelled.")

            sleep(wait_time_between_api_calls)
            elapsed_wait_time += wait_time_between_api_calls

        raise DbtCloudRunTimedOut(
            f"Max attempts reached while checking status of job run with ID: {run_id}"
        )

    def list_run_artifact_links(
        self,
        run_id: int,
    ) -> List[Tuple[str, str]]:
        """
        Lists URLs that can be used to download artifacts from a dbt run

        Args:
            run_id: dbt Cloud job run ID

        Returns:
            List of artifact download URLs

        Raises:
            `DbtCloudListArtifactsFailed`: if API to list dbt artifacts fails

        """
        url = self.dbt_cloud_list_run_artifacts_endpoint_v2(run_id=run_id)
        with self.session.get(url) as list_run_artifacts_response:

            if list_run_artifacts_response.status_code != 200:
                raise DbtCloudListArtifactsFailed(list_run_artifacts_response.reason)

        artifact_paths = list_run_artifacts_response.json().get("data")
        return [
            (
                self.dbt_cloud_get_run_artifact_endpoint_v2(
                    run_id=run_id, path=artifact_path
                ),
                artifact_path,
            )
            for artifact_path in artifact_paths
        ]
